tests/components/sonos/test_init.py

```diff
@@ -132,3 +143,254 @@ async def test_async_poll_manual_hosts_warnings(
+            not in caplog.text
+        )
+        assert (
+            f"Could not get visible Sonos devices from {soco_2.ip_address}"
+            not in caplog.text
+        )
+
+
+async def test_async_poll_manual_hosts_5(
+    hass: HomeAssistant,
+    soco_factory: SoCoMockFactory,
+    entity_registry: er.EntityRegistry,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test both succeed, speakers exist and unavailable, ping succeeds."""
+    soco_1 = soco_factory.cache_mock(MockSoCo(), "10.10.10.1", "Living Room")
+    soco_1.renderingControl = Mock()
+    soco_1.renderingControl.GetVolume = Mock()
+    soco_2 = soco_factory.cache_mock(MockSoCo(), "10.10.10.2", "Bedroom")
+    soco_2.renderingControl = Mock()
+    soco_2.renderingControl.GetVolume = Mock()
+    with patch(
+        "homeassistant.components.sonos.DISCOVERY_INTERVAL"
+    ) as mock_discovery_interval:
+        # Speed up manual discovery interval so second iteration runs sooner
+        mock_discovery_interval.total_seconds = Mock(side_effect=[0.5, 60])
+
+        await _setup_hass(hass)
+
+        assert "media_player.bedroom" in entity_registry.entities
+        assert "media_player.living_room" in entity_registry.entities
+
+        with caplog.at_level(logging.DEBUG):
+            caplog.clear()
+            await asyncio.sleep(1.0)
+            await hass.async_block_till_done()
+            assert "Activity on Living Room" in caplog.text
+            assert "Activity on Bedroom" in caplog.text
+
+
+async def test_async_poll_manual_hosts_6(
+    hass: HomeAssistant,
+    soco_factory: SoCoMockFactory,
+    entity_registry: er.EntityRegistry,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test both succeed, speakers exist and unavailable, pings fail."""
+    soco_1 = soco_factory.cache_mock(MockSoCo(), "10.10.10.1", "Living Room")
+    # Rendering Control Get Volume is what speaker ping calls.
+    soco_1.renderingControl = Mock()
+    soco_1.renderingControl.GetVolume = Mock()
+    soco_1.renderingControl.GetVolume.side_effect = SonosUpdateError()
+    soco_2 = soco_factory.cache_mock(MockSoCo(), "10.10.10.2", "Bedroom")
+    soco_2.renderingControl = Mock()
+    soco_2.renderingControl.GetVolume = Mock()
+    soco_2.renderingControl.GetVolume.side_effect = SonosUpdateError()
+
+    with patch(
+        "homeassistant.components.sonos.DISCOVERY_INTERVAL"
+    ) as mock_discovery_interval:
+        # Speed up manual discovery interval so second iteration runs sooner
+        mock_discovery_interval.total_seconds = Mock(side_effect=[0.5, 60])
+        await _setup_hass(hass)
+
+        assert "media_player.bedroom" in entity_registry.entities
+        assert "media_player.living_room" in entity_registry.entities
+
+        with caplog.at_level(logging.DEBUG):
+            caplog.clear()
+            # Delay for second iteration and other tasks
+            await asyncio.sleep(1.0)
```

Comment:
I'm assuming the sleep is waiting for something happening in another thread.

You could change this to an asyncio.Event or asyncio.Future and do

`await event.wait()`

Than mock the threaded code to do `loop.call_soon_threadsafe(event.set)` so that it unblocks here

Reply:
Yes, there a whole set of tasks that run to create the speaker and managing the subscription failing on the first run of the manual discovery code, those need to complete before the second run of the manual discovery code, then after the second run of the manual discovery code (which occurs 0.5 seconds later), we need to wait for those tasks to also run.  Since there are two ip addresses under test (which was the buggy area); multiple of those need to run.

That test is just trying to make sure that this occurs.

https://github.com/PeteRager/core/blob/f08158ec6db4196e7cdd056c259fe74f911a1da1/homeassistant/components/sonos/__init__.py#L420

```
                    async_dispatcher_send(
                        self.hass,
                        f"{SONOS_SPEAKER_ACTIVITY}-{soco.uid}",
                        "manual zone scan",
                    )
```

But that destination in speaker.py will not call anything in pypi (SoCo) because the speaker has not "cooled down" enough.

Perhaps I'll try to subscribe to that topic directly and have the callback fire the event which will then release the test code?

Reply:
I think it would be better to subscribe to the dispatcher than the sleep.

The dispatcher api isn't likely to change in the future so that should be a quite safe.

Reply:
done